pkg/utils/install/file_source.go

```diff
@@ -0,0 +1,123 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package install
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"slices"
+
+	v1 "github.com/elastic/elastic-agent/pkg/api/v1"
+)
+
+type FileDescriptorSource struct {
+	descriptorFile string
+}
+
+func NewFileDescriptorSource(descriptorFile string) *FileDescriptorSource {
+	return &FileDescriptorSource{descriptorFile: descriptorFile}
+}
+
+func (dp *FileDescriptorSource) AddInstallDesc(desc v1.AgentInstallDesc) (*v1.InstallDescriptor, error) {
+	installDescriptor, err := readInstallMarkerFile(dp.descriptorFile)
+	// not existing or empty files are tolerated, since we would be writing a new descriptor, return any other error
+	if err != nil && !errors.Is(err, os.ErrNotExist) && !errors.Is(err, io.EOF) {
+		return nil, err
+	}
+
+	if installDescriptor == nil {
+		installDescriptor = v1.NewInstallDescriptor()
+	}
+
+	existingInstalls := installDescriptor.AgentInstalls
+	installDescriptor.AgentInstalls = make([]v1.AgentInstallDesc, len(existingInstalls)+1)
+	installDescriptor.AgentInstalls[0] = desc
+	copied := copy(installDescriptor.AgentInstalls[1:], existingInstalls)
+	if copied != len(existingInstalls) {
+		return nil, fmt.Errorf("error adding new install %v to existing installs %v", desc, existingInstalls)
+	}
```

Comment:
WDYT about encapsulating this into a method on `InstallDescriptor`, something like `AddInstall(AgentInstallDesc)`?

Reply:
InstallDescriptor is a data structure defined in `api/v1` package. I'd rather not add methods to that and keep it as a plain data struct